src/Main.hs

```diff
@@ -2,203 +2,178 @@
-render :: IO ()
-render = do
-  putStrLn "Path to save image in?"
-  path <- getLineFixed
-  --- Computes each pixel in //, stores in REPA array of type Array U DIM2 Vec3
-  img <- R.computeUnboxedP generateImgRepa
-  --- Converts to Juicy Pixel ImageRGB16 and saves it to disk as a png
-  (savePngImage path . ImageRGB16 . toImage) img
-
-
---- from A3 sol to clean up user input
-getLineFixed :: IO String
-getLineFixed =
-    do
-        line <- getLine
-        return (fixDel line)
-
-
-fixDel :: String -> String
-fixDel st
-    | '\DEL' `elem` st = fixDel (remDel st)
-    | otherwise = st
-
-
-remDel :: String -> String
-remDel ('\DEL':t) = t
-remDel (_:'\DEL':t) = t
-remDel (_:t) = remDel t
-
-
---- function to generate REPA array of calculated pixel values
---- The D means its delayed, not actually calculated until we call
---- R.computeUnboxedP for parallel, or R.computeUnboxedS for serial
-generateImgRepa :: R.Array D R.DIM2 Vec3
-generateImgRepa = R.fromFunction (Z :. nx :. ny) (calcPixelAt camera world)
-
-
--- | Lazily generate an array of RNGs (one for each AA sample).
-rngs :: Array Int RNG
-rngs = listArray (0, count)
-                 (take count [mkRNG seed | seed <- randoms (mkStdGen 0)])
-    where count = ny * nx * ns + nx * ns + ns - 1
+-- | Render an image with the provided options (and scene)
+render :: Options -> (Camera, [Primitive]) -> IO ()
+render options world = do
+    -- compute each pixel in //, stores in REPA array
+    img <- R.computeUnboxedP (generateImgRepa world options)
+    -- convert to Juicy Pixel ImageRGB16 and save it to disk as a PNG
+    (savePngImage (output options) . ImageRGB16 . toImage) img
 
 
--- | Get the RNG for a given sample.
-getRNG :: Int -> Int -> Int -> RNG
-getRNG x y s = rngs Data.Array.! (y * nx * ns + x * ns + s)
+-- | Generate a REPA array of calculated pixel values.
+-- The 'R.D' means the calculation is delayed until we call 'R.computeUnboxedP'
+-- for parallel evaluation, or 'R.computeUnboxedS' for serial.
+generateImgRepa :: (Camera, [Primitive]) -> Options -> R.Array R.D R.DIM2 Vec3
+generateImgRepa !world !options =
+    let (!camera, !primitives) = world
+        (!nx, !ny) = size camera
+    in R.fromFunction (Z :. nx :. ny) (calcPixelAt camera primitives options)
 
 
---- function to calculate Pixel value at x y in a DIM2 Vec3 array/image plane
-calcPixelAt :: Camera -> [Primitive] -> (Z :. Int :. Int) -> Vec3
-calcPixelAt cam world (Z :. x :. y) =
-    let (red, green, blue) = aaPixel cam world x (ny-y)
+-- | Calculate the pixel value at x y in an image plane.
+calcPixelAt :: Camera -> [Primitive] -> Options -> (Z :. Int :. Int) -> Vec3
+calcPixelAt cam world options (Z :. x :. y) =
+    let (_, ny) = size cam
+        (red, green, blue) = aaSample cam world x (ny - y) options
+    -- convert to an "integer" (will get rounded properly later)
     in (65534.99*(sqrt red), 65534.99*(sqrt green), 65534.99*(sqrt blue))
```

Comment:
@JGAJ Can you explain why you added in the `sqrt` transformation?  I don't understand why it's necessary here

Reply:
Gamma correction, otherwise image is very dark, this hard codes gamma as 2, could refactor and make it more general. (See page 23 in Shirleys tutorial).

Reply:
Ah, I see.  What thre me off is that I'm more used to seeing it as an exponential, and usually people use a gamma of 2.2 since it approximates the [sRGB gamma function better](https://en.wikipedia.org/wiki/SRGB#Theory_of_the_transformation), but this is fine too.